include/stringzilla/stringzilla.hpp

```diff
@@ -458,8 +458,10 @@ class range_matches {
             return temp;
         }
 
-        bool operator!=(iterator const &other) const noexcept { return remaining_.begin() != other.remaining_.begin(); }
-        bool operator==(iterator const &other) const noexcept { return remaining_.begin() == other.remaining_.begin(); }
+        bool operator!=(iterator const &other) const noexcept { return !(*this == other); }
+        bool operator==(iterator const &other) const noexcept {
+            return remaining_.size() == other.remaining_.size() && remaining_.data() == other.remaining_.data();
+        }
```

Comment:
Can you please share, which test is failing? [In C++ you are not expected to compare iterators from different containers](https://stackoverflow.com/a/4664519) and this way we are introducing an extra branch.

Reply:
It fails in `test_search_with_misaligned_repetitions`.

<details>
<summary>Stack trace</summary>
>	stringzilla_test_cpp20.exe!std::_String_view_iterator<std::char_traits<char>>::operator==(const std::_String_view_iterator<std::char_traits<char>> & _Right) Line 1065	C++
 	stringzilla_test_cpp20.exe!ashvardanian::stringzilla::range_matches<std::basic_string_view<char,std::char_traits<char>>,ashvardanian::stringzilla::matcher_find<std::basic_string_view<char,std::char_traits<char>>,ashvardanian::stringzilla::include_overlaps_type>>::iterator::operator!=(const ashvardanian::stringzilla::range_matches<std::basic_string_view<char,std::char_traits<char>>,ashvardanian::stringzilla::matcher_find<std::basic_string_view<char,std::char_traits<char>>,ashvardanian::stringzilla::include_overlaps_type>>::iterator & other) Line 461	C++
 	[External Code]	
 	stringzilla_test_cpp20.exe!test_search_with_misaligned_repetitions<ashvardanian::stringzilla::range_matches<std::basic_string_view<char,std::char_traits<char>>,ashvardanian::stringzilla::matcher_find<std::basic_string_view<char,std::char_traits<char>>,ashvardanian::stringzilla::include_overlaps_type>>,ashvardanian::stringzilla::range_matches<ashvardanian::stringzilla::basic_string_slice<char const>,ashvardanian::stringzilla::matcher_find<ashvardanian::stringzilla::basic_string_slice<char const>,ashvardanian::stringzilla::include_overlaps_type>>>(std::basic_string_view<char,std::char_traits<char>> haystack_pattern, std::basic_string_view<char,std::char_traits<char>> needle_stl, unsigned __int64 misalignment) Line 1011	C++
 	stringzilla_test_cpp20.exe!test_search_with_misaligned_repetitions(std::basic_string_view<char,std::char_traits<char>> haystack_pattern, std::basic_string_view<char,std::char_traits<char>> needle_stl, unsigned __int64 misalignment) Line 1072	C++
 	stringzilla_test_cpp20.exe!test_search_with_misaligned_repetitions(std::basic_string_view<char,std::char_traits<char>> haystack_pattern, std::basic_string_view<char,std::char_traits<char>> needle_stl) Line 1100	C++
 	stringzilla_test_cpp20.exe!test_search_with_misaligned_repetitions() Line 1114	C++
 	stringzilla_test_cpp20.exe!main(int argc, const char * * argv) Line 1319	C++
 	[External Code]	

</details>

Also I think this might actually be allowed because forward_iterators are allowed to be compared is they point to the same container:
```
Comparison between iterators i and j has a defined result if
i and j are iterators to the same underlying sequence, or
both i and j are value-initialized, in which case they compare equal.
```
[Source]()

Reply:
Actually according to [https://stackoverflow.com/questions/52469911/comparing-a-stdstring-view-and-a-substring-string-view] I should just be comparing the data pointers, rather than the data and the length.

Reply:
That makes sense. Maybe worth documenting that behavior or adding a test to run in the pipeline. I've also asked the broader community for advice [on X](https://x.com/ashvardanian/status/1764364205510348847?s=46&t=6AeT2KDQJWzv0tDVl98ZfQ) 🤗